// 4.5/lib/is.js
"use strict";
var typed_colors_1 = require("typed-colors");
var curry_1 = require("@typed/curry");
var helpers_1 = require("./helpers");
exports.is = curry_1.curry(function is(expected, actual) {
    return new Is(expected, actual);
});
var Is = (function () {
    function Is(expected, actual) {
        this.expected = expected;
        this.actual = actual;
    }
    Is.prototype.verify = function (verification) {
        var _a = this, expected = _a.expected, actual = _a.actual;
        if (expected === actual)
            verification.success(actual);
        else
            verification.failure(this.error(expected, actual));
    };
    Is.prototype.error = function (expected, actual) {
        var message = "Not same reference";
        return typed_colors_1.bold(message) + "\n  " + typed_colors_1.green('expected') + ": " + helpers_1.inspect(expected) + "\n    " + typed_colors_1.red('actual') + ": " + helpers_1.inspect(actual) + "\n  ";
    };
    return Is;
}());
exports.Is = Is;
